src/core.c/CompUnit/Repository/Installation.pm6

```diff
@@ -227,145 +227,150 @@ sub MAIN(:$name, :$auth, :$ver, *@, *%) {
-            $lock.unlock;
-            fail "$dist already installed";
-        }
-
-        my $sources-dir   = self!sources-dir;
-        my $resources-dir = self!resources-dir;
-        my $bin-dir       = self!bin-dir;
-        my $is-win        = Rakudo::Internals.IS-WIN;
-
-        self!add-short-name($dist.meta<name>, $dist); # so scripts can find their dist
-
-        my %links; # map name-path to new content address
-        my %provides; # meta data gets added, but the format needs to change to
-                      # only extend the structure, not change it
-
-        # the following 3 `for` loops should be a single loop, but has been
-        # left this way due to impeding precomp changes
-
-        # lib/ source files
-        for $dist.meta<provides>.kv -> $name, $file is copy {
-            # $name is "Inline::Perl5" while $file is "lib/Inline/Perl5.pm6"
-            my $id          = self!file-id(~$name, $dist-id);
-            my $destination = $sources-dir.add($id);
-            my $handle      = $dist.content($file);
-            my $content     = $handle.open(:bin).slurp(:close);
-
-            self!add-short-name($name, $dist, $id,
-              nqp::sha1(nqp::join("\n", nqp::split("\r\n",
-                $content.decode('iso-8859-1')))));
-            %provides{ $name } = ~$file => {
-                :file($id),
-                :time(try $file.IO.modified.Num),
-            };
-            note("Installing {$name} for {$dist.meta<name>}") if $verbose and $name ne $dist.meta<name>;
-            $destination.spurt($content);
-        }
-
-        # bin/ scripts
-        for %files.kv -> $name-path, $file is copy {
-            next unless $name-path.starts-with('bin/');
-            my $name        = $name-path.subst(/^bin\//, '');
-            my $id          = self!file-id(~$file, $dist-id);
-            # wrappers are put in bin/; originals in resources/
-            my $destination = $resources-dir.add($id);
-            my $withoutext  = $name-path.subst(/\.[exe|bat]$/, '');
-            for @script-postfixes -> $be {
-                $.prefix.add("$withoutext$be").IO.spurt:
-                    $raku-wrapper.subst('#name#', $name, :g).subst('#raku#', "rakudo$be");
-                if $is-win {
-                    $.prefix.add("$withoutext$be.bat").IO.spurt:
-                        $windows-wrapper.subst('#raku#', "rakudo$be", :g);
+            my @*MODULES;
+            my $path   = self!writeable-path or die "No writeable path found, $.prefix not writeable";
+            my $lock = $.prefix.add('repo.lock').open(:create, :w);
+            $lock.lock;
+
+            my $version = self!repository-version;
+            self.upgrade-repository unless $version == 2;
+
+            my $dist-id = $dist.id;
+            my $dist-dir = self!dist-dir;
+            if not $force and $dist-dir.add($dist-id) ~~ :e {
+                $lock.unlock;
+                $already-installed = True;
```

Comment:
Is there not a way to keep the original indentation / return style / failure point? I'm guessing its required for the new lock mechanism to work properly, but there are multiple hundred lines of code contained in this lock that anyone else looking here needs to know to e.g. not return early. I find it kind of hard to follow compared to the previous pattern.

Reply:
Unfortunately, not until `LEAVE` handles early `return` correctly. I was very much hesitant about changing this piece and spent hours experimenting with workarounds, but there is none. So, no `return` or `fail` within the protected blocks so far.

Reply:
But what about indirect failures from e.g. some file operation failing?

Reply:
They wouldn't be affected because they'd be handled by the routines where they happen. The problem with `LEAVE` is about blocks only. I guess, it's a kind of optimization for a lexically scoped block to avoid passing through all the stages of exception throwing/catching, which kicks in. As soon as `fail` results in a `Failure` returned it's not a problem anymore.